src/rest-core-types.c

```diff
@@ -98,59 +98,150 @@ void rest_async_response_delete(rest_async_response_t *response)
+    for (i = 0; i < string_length; i++)
+    {
+        if (base64_string[i] == 0x3d)
+        {
+            padding++;
+            if (base64_string[i + 1] == 0x3d)
+            {
+                padding++;
+            }
+            break;
+        }
+    }
+
+    buffer_length = (string_length / 4) * 3 - padding;
+    if (buffer_length <= 0)
+    {
+        return BASE64_ERR_ARG;
+    }
+
+    if (data == NULL)
+    {
+        *length = buffer_length;
+        return BASE64_ERR_NONE;
+    }
+
+    if (*length < buffer_length)
+    {
+        return BASE64_ERR_BUF_SIZE;
+    }
+
+    memset(data, 0, *length);
+
+    uint8_t bits6;
+    const char *pos;
+    for (string_index = 0; string_index < string_length; string_index++)
+    {
+        if (base64_string[string_index] == 0x3d)
+        {
+            break;
+        }
+        if ((pos = strchr(base64_table, base64_string[string_index])) == NULL)
+        {
+            return BASE64_ERR_INV_CHAR;
+        }
+        bits6 = pos - base64_table;
+
+        switch (string_index % 4)
+        {
+        case 0:
+            data[data_index] = data[data_index] | (bits6 << 2);
+            break;
+        case 1:
+            data[data_index] = data[data_index] | (bits6 >> 4);
+            data[data_index + 1] = data[data_index + 1] | (bits6 << 4);
+            break;
+        case 2:
+            data[data_index + 1] = data[data_index + 1] | (bits6 >> 2);
+            data[data_index + 2] = data[data_index + 2] | (bits6 << 6);
+            break;
+        case 3:
+            data[data_index + 2] = data[data_index + 2] | (bits6);
+            data_index = data_index + 3;
+            break;
+        }
+    }
+
+    return BASE64_ERR_NONE;
+}
+
+int base64_encode(const uint8_t *data, size_t length, char *base64_string, size_t *base64_length)
 {
-    size_t buffer_length = ((length + 2) / 3) * 4 + 1; // +1 for null-terminator
-    char *buffer;
     static uint8_t previous_byte;
     int data_index = 0,
         buffer_index = 0;
 
-    buffer = malloc(buffer_length);
-    if (buffer == NULL)
+    if (data == NULL || length <= 0)
     {
-        return NULL;
+        *base64_length = 0;
+        return BASE64_ERR_NONE;
+    }
+
+    int string_length = ((length + 2) / 3) * 4;
+    if (base64_string == NULL)
+    {
+        *base64_length = string_length;
+        return BASE64_ERR_NONE;
+    }
+
+    if (*base64_length < string_length)
+    {
+        *base64_length = string_length;
+        return BASE64_ERR_STR_SIZE;
     }
 
     for (data_index = 0; data_index < length; data_index++)
     {
         switch (data_index % 3)
         {
         case 2:
-            buffer[buffer_index++] = base64_table[
-                                         ((previous_byte & 0x0f) << 2) + ((data[data_index] & 0xc0) >> 6)
-                                     ];
-            buffer[buffer_index++] = base64_table[data[data_index] & 0x3f];
+            base64_string[buffer_index++] = base64_table[
```

Comment:
if you are changing names, then change names properly:
`buffer` to `base64_string` change should cause `buffer_index` variable to be changed to something like `base64_string_index` or `string_index` (I personally would prefer first one, because verbosity usually reduces complexity)